{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "bb9ee442",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1c606d0",
   "metadata": {},
   "source": [
    "# DATA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "b3fa2874",
   "metadata": {},
   "outputs": [],
   "source": [
    "path = \"/mnt/c/Users/aniaf/Projects/PC/GAPP/DATA/AFR_prediction_albicans/10_variants_5sets/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "f777db68",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1348658, 16) (1343984, 19) (1348658, 7) (1348658, 12)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>TYPE</th>\n",
       "      <th>REF_hom</th>\n",
       "      <th>ALT_hom</th>\n",
       "      <th>ALT_het</th>\n",
       "      <th>miss</th>\n",
       "      <th>NGT</th>\n",
       "      <th>ALT_hom_het</th>\n",
       "      <th>fALT_hom_het</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Ca22chr1A_C_albicans_SC5314</td>\n",
       "      <td>62</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>404</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>759</td>\n",
       "      <td>407</td>\n",
       "      <td>3</td>\n",
       "      <td>0.007371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Ca22chr1A_C_albicans_SC5314</td>\n",
       "      <td>64</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>355</td>\n",
       "      <td>32</td>\n",
       "      <td>7</td>\n",
       "      <td>772</td>\n",
       "      <td>394</td>\n",
       "      <td>39</td>\n",
       "      <td>0.098985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Ca22chr1A_C_albicans_SC5314</td>\n",
       "      <td>70</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>373</td>\n",
       "      <td>82</td>\n",
       "      <td>36</td>\n",
       "      <td>675</td>\n",
       "      <td>491</td>\n",
       "      <td>118</td>\n",
       "      <td>0.240326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Ca22chr1A_C_albicans_SC5314</td>\n",
       "      <td>75</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>621</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>544</td>\n",
       "      <td>622</td>\n",
       "      <td>1</td>\n",
       "      <td>0.001608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Ca22chr1A_C_albicans_SC5314</td>\n",
       "      <td>82</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>629</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>536</td>\n",
       "      <td>630</td>\n",
       "      <td>1</td>\n",
       "      <td>0.001587</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        #CHROM  POS REF ALT TYPE  REF_hom  ALT_hom  ALT_het  \\\n",
       "0  Ca22chr1A_C_albicans_SC5314   62   T   C  SNP      404        3        0   \n",
       "1  Ca22chr1A_C_albicans_SC5314   64   A   C  SNP      355       32        7   \n",
       "2  Ca22chr1A_C_albicans_SC5314   70   A   G  SNP      373       82       36   \n",
       "3  Ca22chr1A_C_albicans_SC5314   75   T   A  SNP      621        1        0   \n",
       "4  Ca22chr1A_C_albicans_SC5314   82   G   A  SNP      629        1        0   \n",
       "\n",
       "   miss  NGT  ALT_hom_het  fALT_hom_het  \n",
       "0   759  407            3      0.007371  \n",
       "1   772  394           39      0.098985  \n",
       "2   675  491          118      0.240326  \n",
       "3   544  622            1      0.001608  \n",
       "4   536  630            1      0.001587  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1 = pd.read_csv(path + 'bcftools_FLT2_snpEff_INFO_ANN.tab', sep='\\t', header=0)\n",
    "df1 = df1.rename(columns={'TYPE':'QUAL1','QUAL':'TYPE1'})\n",
    "df1 = df1.rename(columns={'TYPE1':'TYPE','QUAL1':'QUAL'})\n",
    "df2 = pd.read_csv(path + 'bcftools_FLT2_snpEff_gff_table.tab', sep='\\t', header=0)\n",
    "df3 = pd.read_csv('11_07_get_DP.csv', sep=',', header=0)\n",
    "df4 = pd.read_csv('11_08_get_GT.csv', sep=',', header=0)\n",
    "\n",
    "print(df1.shape, df2.shape, df3.shape, df4.shape)\n",
    "df4.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "3985f98f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1348658, 33)\n",
      "(1348658, 35)\n",
      "(1348658, 42)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>TYPE</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>ANN_mut</th>\n",
       "      <th>ANN_type</th>\n",
       "      <th>ANN_effect</th>\n",
       "      <th>ANN_feature</th>\n",
       "      <th>...</th>\n",
       "      <th>feature_annots</th>\n",
       "      <th>DPsum</th>\n",
       "      <th>DPmean</th>\n",
       "      <th>REF_hom</th>\n",
       "      <th>ALT_hom</th>\n",
       "      <th>ALT_het</th>\n",
       "      <th>miss</th>\n",
       "      <th>NGT</th>\n",
       "      <th>ALT_hom_het</th>\n",
       "      <th>fALT_hom_het</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Ca22chr1A_C_albicans_SC5314</td>\n",
       "      <td>62</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>114.9410</td>\n",
       "      <td>C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>C1_00010W_A</td>\n",
       "      <td>...</td>\n",
       "      <td>chromosome=Ca22chr1A_C_albicans_SC5314</td>\n",
       "      <td>15198</td>\n",
       "      <td>13.034305</td>\n",
       "      <td>404</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>759</td>\n",
       "      <td>407</td>\n",
       "      <td>3</td>\n",
       "      <td>0.007371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Ca22chr1A_C_albicans_SC5314</td>\n",
       "      <td>64</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>5092.6100</td>\n",
       "      <td>C</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>C1_00010W_A</td>\n",
       "      <td>...</td>\n",
       "      <td>chromosome=Ca22chr1A_C_albicans_SC5314</td>\n",
       "      <td>18990</td>\n",
       "      <td>16.286449</td>\n",
       "      <td>355</td>\n",
       "      <td>32</td>\n",
       "      <td>7</td>\n",
       "      <td>772</td>\n",
       "      <td>394</td>\n",
       "      <td>39</td>\n",
       "      <td>0.098985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Ca22chr1A_C_albicans_SC5314</td>\n",
       "      <td>70</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>11754.4000</td>\n",
       "      <td>G</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>C1_00010W_A</td>\n",
       "      <td>...</td>\n",
       "      <td>chromosome=Ca22chr1A_C_albicans_SC5314</td>\n",
       "      <td>31654</td>\n",
       "      <td>27.147513</td>\n",
       "      <td>373</td>\n",
       "      <td>82</td>\n",
       "      <td>36</td>\n",
       "      <td>675</td>\n",
       "      <td>491</td>\n",
       "      <td>118</td>\n",
       "      <td>0.240326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Ca22chr1A_C_albicans_SC5314</td>\n",
       "      <td>75</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>126.9440</td>\n",
       "      <td>A</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>C1_00010W_A</td>\n",
       "      <td>...</td>\n",
       "      <td>chromosome=Ca22chr1A_C_albicans_SC5314</td>\n",
       "      <td>41916</td>\n",
       "      <td>35.948542</td>\n",
       "      <td>621</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>544</td>\n",
       "      <td>622</td>\n",
       "      <td>1</td>\n",
       "      <td>0.001608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Ca22chr1A_C_albicans_SC5314</td>\n",
       "      <td>82</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>45.3951</td>\n",
       "      <td>A</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>C1_00010W_A</td>\n",
       "      <td>...</td>\n",
       "      <td>chromosome=Ca22chr1A_C_albicans_SC5314</td>\n",
       "      <td>59833</td>\n",
       "      <td>51.314751</td>\n",
       "      <td>629</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>536</td>\n",
       "      <td>630</td>\n",
       "      <td>1</td>\n",
       "      <td>0.001587</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        #CHROM  POS REF ALT TYPE        QUAL ANN_mut  \\\n",
       "0  Ca22chr1A_C_albicans_SC5314   62   T   C  SNP    114.9410       C   \n",
       "1  Ca22chr1A_C_albicans_SC5314   64   A   C  SNP   5092.6100       C   \n",
       "2  Ca22chr1A_C_albicans_SC5314   70   A   G  SNP  11754.4000       G   \n",
       "3  Ca22chr1A_C_albicans_SC5314   75   T   A  SNP    126.9440       A   \n",
       "4  Ca22chr1A_C_albicans_SC5314   82   G   A  SNP     45.3951       A   \n",
       "\n",
       "                ANN_type ANN_effect  ANN_feature  ...  \\\n",
       "0  upstream_gene_variant   MODIFIER  C1_00010W_A  ...   \n",
       "1  upstream_gene_variant   MODIFIER  C1_00010W_A  ...   \n",
       "2  upstream_gene_variant   MODIFIER  C1_00010W_A  ...   \n",
       "3  upstream_gene_variant   MODIFIER  C1_00010W_A  ...   \n",
       "4  upstream_gene_variant   MODIFIER  C1_00010W_A  ...   \n",
       "\n",
       "                           feature_annots  DPsum     DPmean REF_hom ALT_hom  \\\n",
       "0  chromosome=Ca22chr1A_C_albicans_SC5314  15198  13.034305     404       3   \n",
       "1  chromosome=Ca22chr1A_C_albicans_SC5314  18990  16.286449     355      32   \n",
       "2  chromosome=Ca22chr1A_C_albicans_SC5314  31654  27.147513     373      82   \n",
       "3  chromosome=Ca22chr1A_C_albicans_SC5314  41916  35.948542     621       1   \n",
       "4  chromosome=Ca22chr1A_C_albicans_SC5314  59833  51.314751     629       1   \n",
       "\n",
       "  ALT_het  miss  NGT  ALT_hom_het  fALT_hom_het  \n",
       "0       0   759  407            3      0.007371  \n",
       "1       7   772  394           39      0.098985  \n",
       "2      36   675  491          118      0.240326  \n",
       "3       0   544  622            1      0.001608  \n",
       "4       0   536  630            1      0.001587  \n",
       "\n",
       "[5 rows x 42 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dm0 = pd.merge(df1, df2, on = ['#CHROM','POS'], how = 'left')\n",
    "print(dm0.shape)\n",
    "dm1 = pd.merge(dm0, df3, on = ['#CHROM','POS','REF','ALT','TYPE'], how = 'left')\n",
    "print(dm1.shape)\n",
    "dm = pd.merge(dm1, df4, on = ['#CHROM','POS','REF','ALT','TYPE'], how = 'left')\n",
    "print(dm.shape)\n",
    "dm.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "272a7fca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['#CHROM', 'POS', 'REF', 'ALT', 'TYPE', 'QUAL', 'ANN_mut', 'ANN_type',\n",
       "       'ANN_effect', 'ANN_feature', 'ANN_feature_type', 'ANN_gt', 'ANN_nc',\n",
       "       'ANN_aa', 'ANN_types_all', 'ANN_info_all', 'CDS',\n",
       "       'blocked_reading_frame', 'centromere', 'chromosome', 'exon', 'gene',\n",
       "       'long_terminal_repeat', 'mRNA', 'ncRNA', 'pseudogene', 'rRNA',\n",
       "       'repeat_region', 'retrotransposon', 'snRNA', 'snoRNA', 'tRNA',\n",
       "       'feature_annots', 'DPsum', 'DPmean', 'REF_hom', 'ALT_hom', 'ALT_het',\n",
       "       'miss', 'NGT', 'ALT_hom_het', 'fALT_hom_het'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dm.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "ea06772e",
   "metadata": {},
   "outputs": [],
   "source": [
    "dm.to_csv(path + 'bcftools_FLT2_snpEff_nice_table.tab', sep='\\t', header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5c37264",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mbio",
   "language": "python",
   "name": "mbio"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
